RouteInfo Metadata (#3)
chadhietala/rfcs-1: text/0000-RouteInfo-Metadata.md

- Start Date: 2018-11-02
- RFC PR: (leave this empty)
- Ember Issue: (leave this empty)

# RouteInfo MetaData

## Summary

The RFC introduces the ability to associate application specific metadata with it's corresponding `RouteInfo` object. This also adds a `metadata` field to the `RouteInfo` which will be the return value of `buildRouteInfoMetaData` for it's corresponding `Route`.

```js
// app/route/profile.js
import Route from '@ember/routing/route';

export default Route.extend({

  buildRouteInfoMetaData() {
    return {
      trackingKey: 'page_profile',
      profile: {

>         viewing: this.userId,

Reviewer: I'd _vaguely_ prefer to avoid suggesting folks should store state on the route instance. Maybe this (and the line below) should be using a user service ala `this.user.userId` and `this.user.userLocale`.

        locale: this.userLocale
      }
    }
  }
  // ...
});
```

```js
// app/services/analytics.js
import Service, { inject } from '@ember/service';

export default Service.extend({
  router: inject('router'),
  init() {
    this._super(...arguments);
    this.router.on('routeDidUpdate', (transition) => {
      let { to, from } = transition;

>       let { metadata: fromMeta } = from;

Reviewer: haha, I hate destructuring like this 😸 

I think this is simpler, shorter, and easier:

```js
let fromMeta = from.metadata;
```

      let { metadata: toMeta } = to;
      ga.sendEvent('pageView', {
        from: fromMeta,
        to: toMeta,
        timestamp: Date.now(),
      })
    })
  },
  // ...
});
```

## Motivation

> While the `RouteInfo` object is suffcient in providing developers metadata about the `Route` it self, it is not suffcient in layering on application specific metadata about the `Route`. This metadata could be anything from just a more domain specific name for a `Route` e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.

Reviewer: ```suggestion
While the `RouteInfo` object is sufficient in providing developers metadata about the `Route` itself, it is not sufficient in layering on application specific metadata about the `Route`. This metadata could be anything from just a more domain specific name for a `Route` e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.
```

## Detailed design

### `buildRouteInfoMetaData`

This optional hook is intended to be used as a way of letting the routing system know about any metadata associated with the route. It is called once any time the `Route` is entered.

> #### `Route` Inteface Extension

Reviewer: ```suggestion
#### `Route` Interface Extension
```

```ts
interface Route {
  // ... existing public API
  buildRouteInfoMetaData(): unknown
}
```

### `RouteInfo.metadata`

> The `metadata` optional field on the `RouteInfo` will be populated with the return value of `buildRouteInfoMetaData`. If there is no meta data associated with the `Route`, the `metadata` field will be `null`.

Reviewer: ```suggestion
The `metadata` optional field on the `RouteInfo` will be populated with the return value of `buildRouteInfoMetaData`. If there is no metadata associated with the `Route`, the `metadata` field will be `null`.
```

```ts
interface RouteInfo {
  // ... existing public API
  metadata: Maybe<unknown>;
}
```

This field will also be added to `RouteInfoWithAttributes` as it is just a super-set of `RouteInfo`.

> ## How we teach this

Reviewer: This should probably mention updating both the API docs and the guides. I believe that the super rentals example does deal with tracking which is _exactly_ what this is helpful for...

export default Route.extend({
  user: inject('user'),
  buildRouteInfoMetadata() {
    return {
      trackingKey: 'page_profile',
      user: {
        id: this.user.id,
        type: this.user.type
      }
    }
  }
  // ...
});
```

```js
// app/services/analytics.js
import Service, { inject } from '@ember/service';

export default Service.extend({
  router: inject('router'),
  init() {
    this._super(...arguments);
    this.router.on('routeDidUpdate', (transition) => {
      let { to, from } = transition;
      let { metadata: fromMeta } = from;
      let { metadata: toMeta } = to;
      ga.sendEvent('pageView', {
        from: fromMeta,
        to: toMeta,
        timestamp: Date.now(),
      })
    })
  },
  // ...
});
```

## Motivation

While the `RouteInfo` object is sufficient in providing developers metadata about the `Route` itself, it is not sufficient in layering on application specific metadata about the `Route`. This metadata could be anything from just a more domain specific name for a `Route` e.g. `profile_page` vs `profile.index`, all the way to providing contextual data when the `Route` was visited.

## Detailed design

### `buildRouteInfoMetadata`

> This optional hook is intended to be used as a way of letting the routing system know about any metadata associated with the route. It is called once any time the `Route` is entered.

Reviewer 2: > It is called once any time the `Route` is entered

What exactly does `entered` mean? `routeWillChange`?

Reviewer 2: I think we need to be super specific about when this is called by the system. It needs to be crystal clear that the model hook won't be invoked yet. I can see folks wanting to rely on data fetches in this hook.

Is it possible for this hook to be a performance problem if someone invokes a service that hasn't been instantiated yet?

#### `Route` Interface Extension

```ts
interface Route {
  // ... existing public API
  buildRouteInfoMetadata(): unknown
}
```

### `RouteInfo.metadata`

The `metadata` optional field on the `RouteInfo` will be populated with the return value of `buildRouteInfoMetadata`. If there is no metadata associated with the `Route`, the `metadata` field will be `null`.

```ts
interface RouteInfo {
  // ... existing public API
  metadata: Maybe<unknown>;
}
```

This field will also be added to `RouteInfoWithAttributes` as it is just a super-set of `RouteInfo`.

## How we teach this

> We feel that this a low-level primitive that will allow existing tracking addons to encapsulate. That being said the concept here is pretty simple: What gets returned from `buildRouteInfoMetadata` becomes the value of `RouteInfo.metadata` for that `Route`.

Reviewer 2: I think we need to be prescriptive about appropriate dependencies for the `buildRouteInfoMetadata` hook, since this blocks routing.
